### DISCUSSION TITLE
Clarification on optimistic updates in mutate

### DISCUSSION THREAD

**Discussion by leo-petrucci** [vercel/swr#2426] [Q&A] [answered]
I've just experienced a bug with optimistic updates and I think it stemmed from misunderstanding how the `mutate` API works.

This is my pseudo code:
```js
// pretend like this is fetching data
const { mutate } = useSWR();

// this is called somewhere else in the app
const refresh = (newData) => {
	// if newData exists I want to pass it directly to the cache
	if(newData) {
		await mutate(newData);
	// if not just refetch from endpoint
	} else {
		await mutate();
	}
}
```

Going off the [Mutation documentation page](https://swr.vercel.app/docs/mutation#api):
> `data`: data to update the client cache, or an async function for the remote mutation

From this I understood that whatever new data I wanted to pass to the cache should be passed as the first argument, and **this works for every browser except for Safari**.

The issue I encountered is that in Safari, after calling `mutate`, the data would not be updated.

I tried a couple of things and figured out the code was fixed by changing it to this:
```js
// pretend like this is fetching data
const { mutate } = useSWR();

// this is called somewhere else in the app
const refresh = (newData) => {
	// if newData exists I want to pass it directly to the cache
	if(newData) {
		await mutate(newData, {
		revalidate: false,
		optimisticData: newData,
	});
	// if not just refetch from endpoint
	} else {
		await mutate();
	}
}
```

However I don't understand the difference between `data` and `optimisticData`. In the docs they have very similar descriptions:
> `optimisticData`: data to immediately update the client cache, or a function that receives current data and returns the new client cache data, usually used in optimistic UI.

So I guess my question is, what is the difference between the data passed in the first argument of `mutate` and the data passed within `optimisticData`?

**Comment by koba04** [comment, marked as answer]
One of the differences between `data` and `optimisticData` is that SWR rollbacks `optimisticData` if an error occurred.
You can disable the behavior by setting `rollbackOnError = false`.

In addition to that, SWR updates the cache data from `optimisticData` with the result of the mutation.
You can disable the behavior by setting `populateCache = false`.

**Reply by leo-petrucci** [reply]
So:
- `data` should be the previously available data
- `optimisticData` should be my new data

Should I specify `revalidate: false` when passing `optimisticData`? Will SWR know not to refetch if I give it `optimisticData`?

**Reply by koba04** [reply]
> Should I specify revalidate: false when passing optimisticData?

It depends on if you want to revalidate. If so, that's right.

> Will SWR know not to refetch if I give it optimisticData?

No, that's not relevant with `optimisticData`. It depends on the `revalidate` option.
